### PR TITLE
Implement support for PROXY Protocol

### LINKED ISSUE #174: HAProxy proxy protocol
does anyone ever had the haproxy proxy protocol implementation in aiosmtpd? Thanks in advance.

### PR DESCRIPTION
**Pull request by pepoluan**
## What do these changes do?

Implements support for [the **PROXY** Protocol](https://www.haproxy.com/blog/haproxy/proxy-protocol/), notably the version as [documented in v2.3.0 of the HAProxy program](https://github.com/haproxy/haproxy/blob/v2.3.0/doc/proxy-protocol.txt).

## Are there changes in behavior for the user?

Should not affect existing users.

## Related issue number

Closes #174

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] tox testenvs have been executed in the following environments:
  - [x] Windows 10 (via PyCharm tox runner): `(ALL)`
  - [x] Windows 10 (via PSCore 7.1.1): `(ALL)`
  - [x] Windows 10 (via Cygwin): `qa,py-nocov`
  - [x] Ubuntu 18.04 on WSL 1.0: `(ALL)` + `pypy3-{nocov,cov,diffcov}`
  - [x] FreeBSD 12.2 on VBox: `(ALL)` + `pypy3-{nocov,cov,diffcov}`
  - [x] OpenSUSE Leap 15.2 on VBox:  `(ALL)` + `pypy3-{nocov,cov,diffcov}`
- [x] Documentation reflects the changes
- [x] Add a news fragment into the `NEWS.rst` file

### CODE REVIEW

**Comment by pepoluan** [comment]
> 
> 
> This pull request **introduces 1 alert** when merging [64111e5](https://github.com/aio-libs/aiosmtpd/commit/64111e54eda933ecc88819135369245feb88c455) into [6404ab2](https://github.com/aio-libs/aiosmtpd/commit/6404ab2784beabe483d139f86bc04e901d080980) - [view on LGTM.com](https://lgtm.com/projects/g/aio-libs/aiosmtpd/rev/pr-bbc25d3dee64aec4adc4c63ab51e4ca2dedffc46)
> 
> **new alerts:**
> 
>     * 1 for `__eq__` not overridden when adding attributes

Oh for cryin' out loud!

IT WAS INTENTIONAL. DOESN'T THE "SUPPRESSION" INDICATE THAT ALREADY

**Comment by TimWolla** [comment]
@pepoluan Feel free to ping me for review of this one, once it's good. It appears a proper rebase should be done?

**Comment by pepoluan** [comment]
> 
> 
> @pepoluan Feel free to ping me for review of this one, once it's good. It appears a proper rebase should be done?

Actually, I've been merging back from `master` for every change, so no additional rebase is necessary.

Even if there's a conflict, I prefer to merge from master then pushing back.

If there's no conflict, we can go direct to squamerging and all should be well.

**Comment by TimWolla** [comment]
> Actually, I've been merging back from `master` for every change, so no additional rebase is necessary.

Looking at the diff and the huge number of more than 400 to-be-merged commits it looks like something went wrong. This PR as of right now appears to include unrelated changes.

**Comment by pepoluan** [comment]
> 
> 
> > Actually, I've been merging back from `master` for every change, so no additional rebase is necessary.
> 
> Looking at the diff and the huge number of more than 400 to-be-merged commits it looks like something went wrong. This PR as of right now appears to include unrelated changes.

It was due to all the rebasing that messes up the git history. That is why I don't want to rebase anymore.

(When squamerged it will all play out, no worry about that.)

If you check "Files changed", you will also see that there are lots of ".rst" files being changed. This is because this PR is finalized after #245 was finalized, and I cannot make a clean change to `master` that also applies cleanly to #245 due to the numer & shape of changes. So I merged from #245.

If you want to review the changes, your best bet is to use the "Files changed" tab and either (1) focus only on ".py" files ignoring others, or (2) wait until #245 gets merged, which should prune the number of files changed greatly.

**Comment by waynew** [comment]
FWIW sometimes a rebase will collect more commits than necessary - use `git rebase --interactive` and make sure you only have the commits you're interested in.

**Comment by pepoluan** [comment]
> 
> 
> FWIW sometimes a rebase will collect more commits than necessary - use `git rebase --interactive` and make sure you only have the commits you're interested in.

Thanks! That's a powerful tool, indeed!

I've carried out an interactive rebase, and now it's down to 108 commits.

I'll merge master to the latest state, and do another commitpush.

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:307]: Actual change is new param `proxy_protocol_timeout`. The rest of the diff is indentation (follow Black style)

**Comment by TimWolla** [comment]
> I've carried out an interactive rebase, and now it's down to 108 commits.

Thanks, that's looking much better now. I'll see when I have a bit of time to look deeper into this. Thanks for taking the time to work on this, especially with including all my wishes :smiley:

**Review by TimWolla — COMMENTED** [review]
Some remarks from looking at the diff. Did not yet perform a functional test.

File: aiosmtpd/docs/proxyprotocol.rst
Comment (pepoluan) [reply aiosmtpd/docs/proxyprotocol.rst:29]: Well, I'm using the commit that has been tagged as `v2.3.0`: https://github.com/haproxy/haproxy/tree/v2.3.0

Unless there is plan to shift the tags around, IMHO referring to it as "version 2.3.0" is not wrong.

File: aiosmtpd/docs/proxyprotocol.rst
Comment (pepoluan) [reply aiosmtpd/docs/proxyprotocol.rst:24]: Will do in next commitpush.

Right now I'm writing patches for Unix Socket Controller.

File: aiosmtpd/docs/proxyprotocol.rst
Comment (pepoluan) [reply aiosmtpd/docs/proxyprotocol.rst:143]: It's using the PROXY doc's terms [here](https://github.com/haproxy/haproxy/blob/v2.3.0/doc/proxy-protocol.txt#L207) and [here](https://github.com/haproxy/haproxy/blob/v2.3.0/doc/proxy-protocol.txt#L386).

File: aiosmtpd/docs/proxyprotocol.rst
Comment (pepoluan) [reply aiosmtpd/docs/proxyprotocol.rst:217]: I can do that. Although honestly I don't see a use case for it.

File: aiosmtpd/docs/proxyprotocol.rst
Comment (pepoluan) [reply aiosmtpd/docs/proxyprotocol.rst]: I've changed the CaseStyling for them.

File: aiosmtpd/docs/proxyprotocol.rst
Comment (pepoluan) [reply aiosmtpd/docs/proxyprotocol.rst]: Done.

File: aiosmtpd/docs/proxyprotocol.rst
Comment (TimWolla) [reply aiosmtpd/docs/proxyprotocol.rst:143]: No, I meant the returned values so that one does not have to check the PROXY version before attempting to use the protocol:

TCP4 -> 0x11
TCP6 -> 0x21
UNKNOWN -> 0x00

Comment (pepoluan) [reply aiosmtpd/docs/proxyprotocol.rst:143]: Oh okay.

That's doable.

File: aiosmtpd/docs/proxyprotocol.rst
Comment (TimWolla) [reply aiosmtpd/docs/proxyprotocol.rst:29]: It is not wrong, but also not right. The PROXY spec evolves independently from HAProxy. Two examples:

- When I implemented the unique id TLV I first sent a patch that simply updated the specification and nothing else. Only afterwards I actually implemented client and server support in HAProxy. Sure, this all happened within a single HAProxy release, but that would not have been a requirement.
- Some future extensions might not be useful in context of HAProxy (e.g. hypothetical stuff related to UDP support). They will be part of the PROXY spec, but not be implemented in HAProxy and instead only in other software (e.g. Amazon's LBs).

Thus referring to the PROXY spec compatibility by the HAProxy branch it appeared in might be misleading. That's why I suggested to use the date instead.

Comment (pepoluan) [reply aiosmtpd/docs/proxyprotocol.rst:29]: I much prefer linking to something with a tag, especially a tag that is production level.

The reasoning being that when something is tagged with a "production" tag, it's frozen.

A commit without a tag is, from my point of view, something "ephemeral and may change". Kind of like my commits in this PR; a certain commit may specify such & such behavior, then several commits later I may change my mind and specify a different behavior as I find a drawback to the previous.

This might be subjective, but that's my stand.

File: aiosmtpd/docs/proxyprotocol.rst
Comment (pepoluan) [reply aiosmtpd/docs/proxyprotocol.rst]: Done.

File: aiosmtpd/docs/handlers.rst
Comment (TimWolla) [inline aiosmtpd/docs/handlers.rst:187]: Is this method signature correct? I had no success with a dedicated `proxy_data` header, but could grab the information from `session.proxy_data`.

File: aiosmtpd/proxy_protocol.py
Comment (TimWolla) [inline aiosmtpd/proxy_protocol.py:219]: `val` is a `bytearray` here. It might be my lack of Python expertise, but I think that a `bytes` object might be more useful here, because one usually does not need to handle individual characters, but instead process e.g. the `UNIQUE_ID` as-is.

**Review by TimWolla — COMMENTED** [review]
I did a functional test with our in-house handler and HAProxy in front of it. I tested both with PROXYv1 and PROXYv2, this is working fine.

- I only have the above remark regarding the documentation.
- And the remark regarding `bytearray` vs `bytes`.

File: aiosmtpd/proxy_protocol.py
Comment (pepoluan) [reply aiosmtpd/proxy_protocol.py:219]: `bytes` and `bytearray` are practically the same in Python. They are comparable, slice-able, decode-able, etc.

The only difference is that the latter is mutable, while the former is not.

File: aiosmtpd/docs/handlers.rst
Comment (pepoluan) [reply aiosmtpd/docs/handlers.rst:187]: Oops, my bad. It _should_ be as the documentation indicated, but apparently I forgot to change the implementation.

I'll fix it in the next commitpush.

File: aiosmtpd/proxy_protocol.py
Comment (TimWolla) [reply aiosmtpd/proxy_protocol.py:219]: Understood, thank you for the clarification. Then this is good and I'll have to recheck what I did wrong in my test :smile:

Comment (pepoluan) [reply aiosmtpd/proxy_protocol.py:219]: Just remember you can't directly compare `bytes` to `str`. So the following
    b"123" == "123"
will result in an error, as `str` is "Unicode codepoints". You need to either `decode` the bytes, or `encode` the str, like so:
    (b"123").decode() == "123"
    b"123" == "123".encode()
or if comparing a variable to a literal, don't forget the `b` prefix, like so:
    proxy_data.tlv.UNIQUE_ID = b"some_unique_id"

**Review by TimWolla — APPROVED** [review]
Did not do another functional test after the recent adjustments. I trust that you know what you are doing here. As my remarks have been resolved this LGTM now. Thank you for taking the time to tackle this and also for taking all my suggestions / complaints into account :smiley:

**Comment by pepoluan** [comment]
> 
> 
> Did not do another functional test after the recent adjustments. I trust that you know what you are doing here. As my remarks have been resolved this LGTM now. Thank you for taking the time to tackle this and also for taking all my suggestions / complaints into account 😃

No problem!

It has been a good exercise for me 😉 

This will get merged no later than next Wednesday in 1.4.0.

Currently am focusing on releasing 1.3.1 because 1.3.0 broke people's servers 😓

**Comment by pepoluan** [comment]
Final checks. If all these succeeds, then I'll squamerge.

- [ ] Windows 10 (via PyCharm tox runner): `(ALL)`
- [ ] Windows 10 (via PSCore 7.1.2): `(ALL)`
- [ ] Windows 10 (via Cygwin): `(ALL)`
- [ ] Ubuntu 18.04 on WSL 1.0: `(ALL)` + `static` + `pypy3-{nocov,cov,diffcov}`
- [ ] FreeBSD 12,2 on VBox: `(ALL)` + `static` + `pypy3-{nocov,cov,diffcov}`
- [ ] OpenSUSE Leap 15.2 on VBox: `(ALL)` + `static` + `pypy3-{nocov,cov,diffcov}`

**_This post is undergoing edits._**

**Pull request merged**
